[REVERTED] Beam Job for populating new models used in Contributor Admin Dashboard: GSoC M1.4

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This fixes part of #17440 .
2. This PR does the following: Beam Job for populating new models used in contributor admin dashboard using existing stats models and general suggestion models.

This beam Job uses below mentioned models:-

- GeneralSuggestionModel
- TranslationContributionStatsModel
- TranslationReviewStatsModel
- QuestionContributionStatsModel
- QuestionReviewStatsModel

Above stats models are also keyed on topic but the stats we are planning to display in contributor admin dashboard isn't related to topics so we have need to introduce these new storage models. The discussion about this and the decision table can be found in the proposal linked above.

Below mentioned models will be populated:-
- TranslationSubmitterTotalContributionStatsModel
- TranslationReviewerTotalContributionStatsModel
- QuestionSubmitterTotalContributionStatsModel
- QuestionReviewerTotalContributionStatsModel

All models are present in core/storage/suggestion/gae_models.py

The plan used to write this beam job is documented [here](https://docs.google.com/document/d/19lCEYQUgV7_DwIK_0rz3zslRHX2qKOHn-t9Twpi0qu0/edit#heading=h.q8pchv9rnpz8).
The need for these models are also documented in the same document [GSoC Proposal](https://docs.google.com/document/d/19lCEYQUgV7_DwIK_0rz3zslRHX2qKOHn-t9Twpi0qu0/edit)


New models populated were introduced in https://github.com/oppia/oppia/pull/18340 PR.


After initial load is populated using the beam Job, new services to update these models will be introduced in the PR GSoC M1.5 i.e. the PR after this.


## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

Same data is used in writing unit tests as well. So if you want to have a look at the data you can look at test file of the beam job.

[Screencast from 18-06-23 02:54:15 PM IST.webm](https://github.com/oppia/oppia/assets/91470808/6f441acb-86dc-4236-9dd4-605116ab50c6)

New Ouput 
![image](https://github.com/oppia/oppia/assets/91470808/a23fcfea-bc85-4041-8f89-cbe4ff76ad2f)

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
initial commit

kshitij01042002 committed:
initial commit, beam jobs added

kshitij01042002 committed:
added tests and docstrings

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 commented:
Hi reviewers, I am working on the testing instruction doc and will soon provide it in the PR description.Till then can you PTAL at other things?

Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

kshitij01042002 committed:
added test to backend_test_shards

kshitij01042002 committed:
Merge branch 'gsoc_M1.4' of https://github.com/kshitij01042002/oppia into gsoc_M1.4

kshitij01042002 committed:
fixed typo in backend_test_shards.json

lkbhitesh07 reviewed (changes requested):
Hey @kshitij01042002, thanks for the changes. Few points:
- Can you please add more things to the PR description? Can you also name down the model names that you are introducing and maybe a small intro about them like what do we need it for..etc.
- I'm assuming that you are introducing some new models to the datastore. How will the new entries to these models will take place? How are the updates going to take place? I think you would also need to introduce models in the gae_models file.
- Can you please also add audit job as this job makes changes to the datastore(I mentioned about this in the mail as well)
- How will you verify that the changes done are correct?

Thanks

kshitij01042002 committed:
added entry in valid.py

kshitij01042002 committed:
added audit job

kshitij01042002 committed:
added Audit job

kshitij01042002 commented:
Hi @lkbhitesh07 thanks for the review.

I have added more information to the PR description PTAL.
I have also mentioned the link to the PR which introduces the new gae_models we are creating.
Added the Audit Job as well.

I am not sure how can I verify that the changes made are correct. The unit tests I have written checks the new models created are the stats/transformed correctly, other than that how can I check?

Thanks.

DubeySandeep commented on core/jobs/batch_jobs/contributor_admin_stats_jobs.py:
```diff
@@ -0,0 +1,1236 @@
+            return question_review_stats_models
+
+
+class AuditGenerateContributorAdminStatsJob(base_jobs.JobBase):
```
Can we avoid adding all the duplicate codes here?

Possible way:

```
class ActualJob:
    DATASTORE_UPDATES_ALLOWED = True
    def run(self):
        .....
        .....
        if DATASTORE_UPDATES_ALLOWED:
            <Steps calling ndb_io.PutModels()>
            
class AuditActualJob(ActualJob):
    DATASTORE_UPDATES_ALLOWED = False
```

@vojtechjelinek Do we have any official way to write audit jobs without duplicate code?
Note: The above-suggested way is a little unsafe as we have to control put calls manually, can be made better if we define (wrapper) self.PutModels method in the Job base class and make ndb_io.PutModels() there as required.

kshitij01042002 replied:
Hi @DubeySandeep thanks for your suggestion but I am little confused on how it should be implemented cause the run method of ActualJob takes `self` as param. Can you please clarify a little bit more?

DubeySandeep replied:
@kshitij01042002 Sorry for the confusion, the above code is for reference and doesn't have the correct code. You can check [this existing job for help](https://github.com/oppia/oppia/blob/5957b9ddc0b58aa61477a3685195b39ac015dfe4/core/jobs/batch_jobs/translation_migration_jobs.py#L195).

kshitij01042002 replied:
Thanks @DubeySandeep for the reference made the suggested changes PTAL.

vojtechjelinek replied:
@DubeySandeep I already responded somewhere else to this, there isn't really a standardized way to do this, it needs to be thought through.

DubeySandeep commented on scripts/linters/test_files/valid.py:
```diff
@@ -45,6 +45,7 @@
     rejecting_suggestion_for_invalid_content_ids_jobs)
 from core.jobs.batch_jobs import (                                   # pylint: disable=unused-import  # isort: skip
     remove_profile_picture_data_url_field_jobs)
+from core.jobs.batch_jobs import contributor_admin_stats_jobs        # pylint: disable=unused-import  # isort: skip
```
Just curious, why do we need this change?

kshitij01042002 replied:
If we don't do this changes,https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter_test.py#L67 this test fails. It requires for every job in batch_jobs/ to be imported in this file.

DubeySandeep replied:
That's weird, thanks for the explanation!

gp201 replied:
@kshitij01042002 Can you point to the documentation where this is mentioned?

kshitij01042002 replied:
Actually I didn't find any documentation related to this but after pushing the changes [this](https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter_test.py#L67) test was failing. On debugging and analyzing the test, also taking reference of other beam job PRs. I figured the above mentioned behaviour.

gp201 replied:
Hmm, that's weird. Can you add a comment in that file saying that every batch_jobs should be added to that file and give the reason for it also?

DubeySandeep commented on core/jobs/batch_jobs/contributor_admin_stats_jobs.py:
```diff
@@ -0,0 +1,1236 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
```
(minor nit) 2023*

kshitij01042002 replied:
Done. Thanks.

DubeySandeep reviewed (changes requested):
@kshitij01042002 Thanks for the PR, I took a quick pass and left a few comments, PTAL!

JayVivarekar commented on core/jobs/batch_jobs/contributor_admin_stats_jobs_test.py:
```diff
@@ -0,0 +1,1260 @@
+            language_code='hi',
+            edited_by_reviewer=False,
+            created_on=datetime.datetime(2023, 2, 2))
+
```
Please add a test for empty storage, which ensures the output is empty for a job run on an empty storage. (Refer to other job test files)

kshitij01042002 replied:
Added PTAL. Thanks for the review.

JayVivarekar reviewed (changes requested):
Thanks @kshitij01042002! Looks fairly good to me. Left a comment, PTAL.

kshitij01042002 committed:
added tests for empty models

kshitij01042002 committed:
removed duplicate code from audit job

kshitij01042002 committed:
fixed lint errors

kshitij01042002 commented:
Hi @JayVivarekar @DubeySandeep @lkbhitesh07 thanks for the review. I have addressed your comments. PTAL.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

qinghaoyang reviewed (approved):
@kshitij01042002 LGTM

DubeySandeep commented on core/jobs/batch_jobs/contributor_admin_stats_jobs_test.py:
```diff
@@ -0,0 +1,1266 @@
+        self.assertEqual(100, len(question_model.recent_review_outcomes))
+
+
+class AuditGenerateContributorAdminStatsJobTests(job_test_utils.JobTestBase):
```
Instead of duplicate setup code can we do the following:

```
class ContributorDashboardTest(job_test_utils.JobTestBase):

    def setUp(self):
        <all-setup>
        
class GenerateContributorAdminStatsJobTests(ContributorDashboardTest):

    def <test-job-logs-for-one-case>
    def <test-job-logs-for-another-case>
    def <test-job-creates/updates-datastore-models>


class AuditGenerateContributorAdminStatsJobTests(ContributorDashboardTest):

    def <test-job-doesnot-creates/updates-datastore-models>

```

kshitij01042002 replied:
Thanks for the suggestion, that does saves lot of duplicate code. Pushed new changes. PTAL.

lkbhitesh07 reviewed (changes requested):
Thanks for the changes, Just one small nit. Could you please also output the name of the model for which we are getting the results? Currently, the job output is like a bunch of "SUCCESS" messages and in case anything gets wrong we will not be sure which model got affected.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

kshitij01042002 committed:
Added message to success Pcollection

kshitij01042002 committed:
Merge branch 'gsoc_M1.4' of https://github.com/kshitij01042002/oppia into gsoc_M1.4

kshitij01042002 committed:
fixed lint errors

kshitij01042002 commented:
Hi @lkbhitesh07, thanks for the suggestion, it sg. Added the names of the models in front of the message. PTAL.

[Pointer](https://github.com/oppia/oppia/pull/18406/files#diff-540ef111c5807b3b0af14252c442cf0f18db3d9623d8c5b60cf97ad32a0524d4R233) to the changes.

lkbhitesh07 reviewed (approved):
Thanks for the changes. Great work!
Just for future reference and future reviewers on this PR also paste the screenshot of the new changes you have made.
Thanks

kshitij01042002 committed:
fixed tests

kshitij01042002 commented:
> Thanks for the changes. Great work! Just for future reference and future reviewers on this PR also paste the screenshot of the new changes you have made. Thanks

Sure, added screenshot in the PR description. Thankyou for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

DubeySandeep reviewed (changes requested):
@kshitij01042002 I've reviewed the general structure of the jobs (not the internal details) and left one comment for the test structure. Other than that LGTM!

kshitij01042002 committed:
changed tests setup

JayVivarekar reviewed (approved):
Thanks @kshitij01042002! LGTM

gp201 reviewed (changes requested):
Thanks, @kshitij01042002 ! I left some comments. PTAL

kshitij01042002 commented:
@gp201 I have replied to your comment. PTAL. Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

kshitij01042002 committed:
added more information in valid.py

kshitij01042002 commented:
Hi @gp201 I have added the comment and reasoning best according to my understanding. PTAL. Thanks.

gp201 commented on scripts/linters/test_files/valid.py:
```diff
@@ -16,6 +16,10 @@
 python_linter_test.py. This file contain valid python syntax.
+Every job in core/jobs/batch_jobs should be imported in this file.
+If we don't import
+https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter_test.py#L67
```
Can you change this to the test name?

gp201 reviewed (changes requested):
1 final nit.

kshitij01042002 committed:
removed link and added the test name

kshitij01042002 commented:
Hi @gp201 made the requested changes. PTAL. Thanks for the review.

gp201 commented on scripts/linters/test_files/valid.py:
```diff
@@ -16,6 +16,8 @@
+Every job in core/jobs/batch_jobs should be imported in this file.
+If we don't import test_valid_job_imports fails
 """
 
```
```suggestion
"""

# Note: Make sure to import every job from the core/jobs/batch_jobs directory into this file.
# Failure to import will cause the test_valid_job_imports linter test to fail.
```

gp201 reviewed (changes requested):
1 final nit

kshitij01042002 committed:
Update scripts/linters/test_files/valid.py

Co-authored-by: Praneeth <praneethg2001@gmail.com>

kshitij01042002 committed:
fixed pylint error

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4

kshitij01042002 commented:
Hi @gp201 updated the final nit. PTAL. Thanks.

gp201 reviewed (approved):
LGTM

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.4
